src/ast/passes/semantic_analyser.cpp

```diff
@@ -37,9 +37,53 @@ static const std::map<std::string, std::tuple<size_t, bool>> &getIntcasts()
   return intcasts;
 }
 
+template <typename T>
+static std::pair<T, T> getIntTypeRange(const SizedType &ty)
+{
+  assert(ty.IsIntegerTy());
+  auto size = ty.GetSize();
+  if (size == 1) {
+    if (ty.IsSigned()) {
+      return { CHAR_MIN, CHAR_MAX };
+    } else {
+      return { 0, UCHAR_MAX };
+    }
+  } else if (size == 2) {
+    if (ty.IsSigned()) {
+      return { SHRT_MIN, SHRT_MAX };
+    } else {
+      return { 0, USHRT_MAX };
+    }
+  } else if (size == 4) {
+    if (ty.IsSigned()) {
+      return { INT_MIN, INT_MAX };
+    } else {
+      return { 0, UINT_MAX };
+    }
+  } else if (size == 8) {
+    if (ty.IsSigned()) {
+      return { LONG_MIN, LONG_MAX };
+    } else {
+      return { 0, ULONG_MAX };
+    }
+  }
+  LOG(BUG) << "Unrecognized int type size: " << size;
+  return { 0, 0 };
+}
+
 void SemanticAnalyser::visit(Integer &integer)
 {
-  integer.type = CreateInt64();
+  if (integer.is_negative) {
+    integer.type = CreateInt64();
+  } else {
+    // Default to signed unless the original value is too large
+    uint64_t val = static_cast<uint64_t>(integer.n);
+    if (val > LONG_MAX) {
```

Comment:
To get consistent behaviour across platforms:
```suggestion
    if (val > std::numeric_limits<int64_t>::max()) {
```

`LONG_MAX` depends on `long`, which can be either 32-bit or 64-bit.